controllers/kyma_controller.go

```diff
@@ -429,3 +430,12 @@ func (r *KymaReconciler) deleteModule(ctx context.Context, moduleStatus *v1beta1
 	manifest.SetName(moduleStatus.Manifest.GetName())
 	return r.Delete(ctx, &manifest, &client.DeleteOptions{})
 }
+
+// RecordKymaStatusMetrics updates prometheus metrics defined to track changes to the Kyma status
+func (r *KymaReconciler) RecordKymaStatusMetrics(kyma *v1beta1.Kyma) {
+	//TODO: where to take "instanceId" and "shoot" from?
+	err := metrics.RecordKymaStatus(kyma.Name, string(kyma.Status.State), "instanceId", "shoot")
```

Comment:
In my opinion the only thing the whole metric setting code needs to boil down to is e.g. if Kyma is in state 'Ready':
```
statusMetric.With(prometheus.Labels{"kyma_name":kyma.Name, "state":"Ready"}).Set(1)
statusMetric.With(prometheus.Labels{"kyma_name":kyma.Name, "state":"Error"}).Set(0)
statusMetric.With(prometheus.Labels{"kyma_name":kyma.Name, "state":"Processing"}).Set(0)
statusMetric.With(prometheus.Labels{"kyma_name":kyma.Name, "state":"Deleting"}).Set(0)
```
so basically every time a Kyma gets reconciled you create new values, since it's important **when** a Kyma turned from Ready -> Processing -> Error for example.

Have you tried this or am I missing some requirement?

Reply:
@lindnerby I think this is exactly what I am doing. Take a look [here](https://github.com/kyma-project/lifecycle-manager/pull/418/files#diff-4bc91a74bc47b8467cdcd25a7e3fe1651dbe44907f478377eca31b982f444f23R33)

Reply:
@Tomasz-Smelcerz-SAP Thanks for the sync. As discussed we:
* should only set labels `name` and `state`
* should probably align with controller-runtime's metric handling implementation and have an init for default values
* as well as may opt into using the previously mentioned method `With(Labels)` over `GetMetricWith(Labels)` to panic on dev errors to _fail-fast_ e.g. for mismatching label cardinality. But only after verifying that the errors are purely about label format, structure, etc. and can not stem from dynamic values from the CR, since controller-runtime is doing that as well. Don't know if we want to risk create only a warning event at runtime if a developer is making an error at some point in the future when e.g. adding new labels, new metrics, etc. Also no metrics would be updated anymore until someone notices that warning events are created.
* maybe for readability you could just iterate over all available statuses and set to 1 if Kyma status matches and to 0 if not

Reply:
@lindnerby The errors are only related to number (and names) of labels, so they don't depend on runtime values. It means we can use a "fail-fast" approach, as the controller-runtime does. I've also introduced similar metrics for the modules statuses. Please take a look